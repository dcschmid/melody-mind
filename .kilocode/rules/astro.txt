Astro-Specific Guidelines

The following guidelines apply specifically to development with Astro and should be observed for all Astro-related tasks.

1. Component Architecture Best Practices

   - Content-based Component Division: Divide components based on content logic, not just visual aspects.
   
   - Astro Components vs. Native JavaScript: Use Astro components (.astro) for static UI elements and native JavaScript/TypeScript for interactive elements that require client-side functionality. This approach reduces bundle size, eliminates framework-specific dependencies, improves performance, and simplifies the development process by leveraging standard web technologies rather than external frameworks.
   
   - Effective Slot System Usage: Utilize Astro's named and unnamed slots for flexible component composition and improved reusability.
   
   - Component Islands: Implement the "Islands Architecture" concept by isolating interactive components and rendering the rest of the page statically.
   
   - Props Validation: Define prop types with TypeScript or JSDoc for better developer experience and error prevention.
   
   - Internationalisierung von Komponententexten: Alle statischen Texte in neuen Komponenten müssen in der 'src/i18n/ui.ts' Datei für jede unterstützte Sprache eingetragen werden. Dies ermöglicht eine konsistente mehrsprachige Unterstützung und vereinfacht die Wartung von Übersetzungen. Verwende beschreibende Schlüssel im Format 'componentname.textbeschreibung' (z.B. 'navbar.home' oder 'footer.copyright').

2. Optimal Use of Static and Dynamic Rendering

   - SSG vs. SSR: Use static rendering (SSG) as the default for maximum performance and caching benefits. Use server-side rendering (SSR) only for pages that require current data or user-specific content.
   
   - Hybrid Rendering: Combine static and dynamic rendering strategies within the same application based on the requirements of each route.
   
   - Incremental Rendering: Use Astro's On-Demand Rendering (ODR) for large websites to optimize build times.
   
   - Prerendering: Identify pages that can be pre-rendered to improve loading times and reduce server load.
   
   - Dynamic Routes: Implement dynamic routes with getStaticPaths() for optimal performance with data-driven pages.

3. Framework Integration Guidelines

   - Minimal Framework Usage: Integrate frameworks (React, Vue, Svelte, etc.) only where interactivity is needed to minimize bundle sizes.
   
   - Client Directives: Use the appropriate client:* directives (client:load, client:idle, client:visible, etc.) based on the component's loading timing.
   
   - Shared State: Implement cross-framework state management with Nano Stores or similar lightweight solutions.
   
   - Hydration Strategies: Choose the right hydration strategy based on component priority and visibility.
   
   - Framework Isolation: Keep different framework components isolated to avoid conflicts and unexpected behavior.

4. Performance Best Practices for Astro

   - Asset Optimization: Use Astro's integrated image optimization with <Image /> and <Picture /> components for optimal loading times.
   
   - CSS Optimization: Use Astro's CSS scoping and bundling for minimal CSS payloads.
   
   - Lazy Loading: Implement lazy loading for images, components, and scripts that are not immediately visible.
   
   - Prefetching: Use <link rel="prefetch"> for important navigation routes to improve perceived loading time.
   
   - Caching Strategies: Implement effective caching strategies with Astro's integrated functions and Service Workers.
   
   - Bundle Analysis: Monitor and optimize bundle sizes regularly with tools like astro-bundle-analyzer.

5. Tailwind Integration in Astro

   - Setup and Configuration: Install @astrojs/tailwind integration and configure tailwind.config.mjs to work with Astro's file structure. Use the applyBaseStyles option to control whether Tailwind's base styles are automatically included.
   
   - Component Styling Strategies: Leverage Astro's component-scoped CSS alongside Tailwind utility classes. For complex components, consider using a hybrid approach where layout is handled by Tailwind and component-specific styling uses Astro's scoped CSS.
   
   - Optimization Techniques: Enable JIT mode for faster builds and smaller CSS output. Use the safelist option for dynamic class names that might be missed by the purge process. Configure content paths properly to ensure all template files are scanned.
   
   - Common Patterns and Solutions: Use @apply directives in Astro's <style> blocks for repeated utility combinations. Create consistent design tokens by extending Tailwind's theme in the configuration. Implement dark mode using Tailwind's dark: variant with Astro's client-side scripts for toggling.
   
   - Integration with Existing Design Systems: Map design system tokens to Tailwind's theme configuration. Create component abstractions that encapsulate design system patterns while using Tailwind under the hood. Use Tailwind's plugin system to extend functionality with custom utilities that match your design system.

6. SEO Best Practices for Astro

   - SEO.astro Component: Use the project's SEO.astro component for consistent and comprehensive SEO implementation across all pages. This component centralizes SEO management and ensures best practices are followed automatically. The component handles meta tags, structured data, internationalization, performance optimizations, and accessibility features.
   
   ```astro
   <SEO
     title="Rock Music Quiz"
     description="Test your knowledge of classic rock hits"
     type="game"
   />
   ```
   
   - SEO.astro Features: The component provides extensive functionality including:
     * Dynamic meta tag generation based on page content
     * Structured data (JSON-LD) with schema.org standards for different content types (website, article, music, game)
     * Internationalization support with automatic hreflang tag generation
     * Performance optimizations with resource hints and critical resource preloading
     * Accessibility information in structured data
     * Mobile app integration meta tags
     * Music-specific OpenGraph tags for music content
   
   - Extending SEO.astro: When adding new features to the SEO component, ensure they:
     * Maintain TypeScript type safety through the Props interface
     * Follow schema.org standards for structured data
     * Support all content types (website, article, music, game)
     * Include appropriate accessibility information
     * Integrate with the existing layout components
   
   - Meta Tags Implementation: Implement comprehensive meta tags in the <head> section using the SEO.astro component. Include title, description, canonical URLs, Open Graph tags, Twitter Card tags, and language attributes. Use dynamic meta tags for content-specific pages.
   
   - Structured Data: Implement JSON-LD structured data for rich search results through the SEO.astro component. Use appropriate schemas from schema.org based on content type (Article, Product, FAQ, etc.). Validate structured data using Google's Rich Results Test.
   
   - Semantic HTML: Use proper semantic HTML elements (header, nav, main, article, section, aside, footer) to improve content structure and accessibility. Implement proper heading hierarchy (h1-h6) with a single h1 per page. Use descriptive anchor text for links.
   
   - Performance Optimization: Optimize Core Web Vitals metrics (LCP, FID/INP, CLS) using Astro's built-in performance features and the SEO.astro component's resource hints. Implement responsive images with srcset and sizes attributes. Use font display strategies to prevent layout shifts. Minimize JavaScript and CSS to reduce blocking time.
   
   - Content Structure: Organize content with clear information architecture. Implement breadcrumb navigation for deep site structures. Create XML sitemaps with prioritization. Implement proper internal linking strategies. Use descriptive URLs with relevant keywords.
   
   - Social Media Optimization: Implement platform-specific meta tags for optimal social sharing through the SEO.astro component. Create properly sized and optimized social sharing images. Use Astro's Image component for social media preview images to ensure optimal loading.
   
   - Integration with Layout Components: Ensure the SEO.astro component works seamlessly with other layout components like SkipLink.astro and Navigation.astro. The SEO component should be included in the Layout.astro file to ensure it's present on all pages.

7. WCAG AAA Compliance Requirements

   - Accessibility Standards: Implement WCAG 2.2 AAA level compliance across all components and pages. Conduct regular automated and manual accessibility audits. Document accessibility features and maintain an accessibility statement.
   
   - Color Contrast Requirements: Ensure text contrast ratio of at least 7:1 for normal text and 4.5:1 for large text. Avoid using color alone to convey information. Provide high contrast modes with contrast ratios of at least 10:1. Test all color combinations in both light and dark themes.
   
   - Keyboard Navigation: Ensure all interactive elements are accessible via keyboard with visible focus indicators. Implement logical tab order matching visual layout. Provide keyboard shortcuts for common actions with the ability to remap or disable them. Ensure no keyboard traps exist.
   
   - Screen Reader Optimization: Use ARIA attributes appropriately when HTML semantics are insufficient. Implement proper landmark roles. Provide descriptive alt text for images. Ensure dynamic content changes are announced to screen readers. Test with multiple screen readers (NVDA, JAWS, VoiceOver).
   
   - Form Accessibility: Implement proper label associations with form controls. Provide clear error messages and validation feedback. Allow sufficient time for form completion with the option to extend time limits. Ensure form controls have sufficient size for touch interaction (at least 44x44px).
   
   - Media Accessibility: Provide captions, transcripts, and audio descriptions for all media content. Ensure media players have accessible controls. Avoid auto-playing media or provide mechanisms to disable autoplay. Ensure no content flashes more than three times per second.

Context7 MCP Usage for Astro Development

The following guidelines describe how the Context7 MCP tool can be effectively used for Astro development.

1. Retrieving Current Astro Documentation

   - Current API Reference: Use Context7 to retrieve the latest Astro API documentation, especially for new features and breaking changes in Astro 4.x and higher.
   
   - Integration Documentation: Use Context7 to obtain specific information about Astro integrations (e.g., @astrojs/react, @astrojs/tailwind).
   
   - Migration Guides: Access current migration guides when projects need to be updated to newer Astro versions.
   
   - Code Examples: Retrieve relevant code examples for specific Astro functions to implement best practices.

2. Access to Best Practices and Patterns

   - Architecture Patterns: Use Context7 to identify proven architecture patterns for different types of Astro projects (blogs, e-commerce, documentation sites, etc.).
   
   - Performance Optimizations: Access current performance optimization strategies specifically developed for Astro.
   
   - SEO Best Practices: Use Context7 to retrieve SEO best practices for Astro websites, including structured data and meta tags.
   
   - Accessibility Patterns: Implement accessible components and patterns based on Context7 recommendations.
   
   - Internationalization Strategies: Gain access to proven i18n implementations for multilingual Astro websites.

3. Debugging Support

   - Error Diagnosis: Use Context7 to identify and fix common Astro-specific errors.
   
   - Build Optimization: Get help diagnosing and solving build performance problems.
   
   - Hydration Errors: Identify and fix hydration errors in framework components with Context7 support.
   
   - SSR Troubleshooting: Solve server-side rendering problems with specific debugging techniques.
   
   - Deployment Problems: Diagnose and fix problems with various hosting platforms (Vercel, Netlify, Cloudflare Pages, etc.).

When using Context7 for Astro development, the following workflow should always be followed:

1. Identify the specific Astro-related problem or requirement
2. Use the resolve-library-id tool with "astro" or specific Astro packages
3. Retrieve the relevant documentation with get-library-docs
4. Apply the obtained information in the context of the current project
5. Validate the implementation against current best practices