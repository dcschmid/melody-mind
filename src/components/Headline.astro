---
/**
 * Headline Component
 *
 * A versatile and accessible heading component that supports different heading levels,
 * custom styling, and proper semantic hierarchy. Follows WCAG AAA accessibility standards
 * and performance best practices.
 *
 * @component
 */
interface Props {
  /** Text content of the heading (can also use slot content instead) */
  title?: string;
  /** HTML heading level (h1-h6) for proper document hierarchy */
  level?: "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
  /** Additional CSS classes to apply */
  className?: string;
  /** Optional ID for direct linking and document structure */
  id?: string;
  /** Adds tabindex="-1" to make the heading programmatically focusable when needed
   * (useful for skip navigation or focus management)
   */
  focusable?: boolean;
  /** Optional ARIA label for improved screen reader context */
  ariaLabel?: string;
}

const {
  title,
  level = "h1",
  className = "text-2xl md:text-4xl",
  id,
  focusable = false,
  ariaLabel,
} = Astro.props;

// Dynamic heading tag based on level prop
const Tag = level;

/**
 * Compute contrast-compliant text color
 * Sky-500 has been verified to meet WCAG AAA 7:1 contrast ratio
 * against the dark background of MelodyMind
 */
const headlineClasses = `${className} font-bold tracking-wide mb-6 text-balance text-sky-500 break-words`;

// Focusable attributes for accessibility when needed (skip navigation)
const focusableAttrs = focusable
  ? {
      tabindex: "-1",
      class: `${headlineClasses} focus:outline-none focus:ring-2 focus:ring-sky-500 focus:ring-offset-2 focus:ring-offset-zinc-900`,
    }
  : { class: headlineClasses };
---

<Tag id={id} aria-label={ariaLabel} {...focusableAttrs}>
  {title || <slot />}
</Tag>
