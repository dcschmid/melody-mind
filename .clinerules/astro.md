# Astro Rules for MelodyMind

## Component Structure

- Follow the Astro component structure with frontmatter script and template
- Use Astro slots for flexible component composition
- Leverage Astro's built-in partial hydration with directives when needed
- Keep component logic in the frontmatter section
- Extract complex logic to utility files

## Page Organization

- Use a consistent page structure with proper SEO metadata
- Implement layouts for consistent UI across pages
- Use dynamic routes for game categories and genres
- Implement proper navigation between pages

## Data Flow

- Use Astro.props for component data
- Implement server-side data fetching where appropriate
- Use Astro.glob() for working with collections of files
- Work with Astro DB for persistent data storage

## Example Component Pattern

```astro
---
// Script section (frontmatter)
import { GameCard } from "../components/GameCard";
import type { GameCategory } from "../types";

interface Props {
  categories: GameCategory[];
  title: string;
}

const { categories, title } = Astro.props;
---

<!-- Template section -->
<section class="game-categories">
  <h2>{title}</h2>
  <div class="card-grid">
    {
      categories.map((category) => (
        <GameCard
          title={category.title}
          difficulty={category.difficulty}
          genre={category.genre}
          client:load
        />
      ))
    }
  </div>
</section>

<style>
  .card-grid {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    gap: 1rem;
  }
</style>
```

## Performance Considerations

- Use client:load, client:idle, or client:visible appropriately
- Optimize images with Astro's built-in image optimization
- Implement lazy loading for below-the-fold content
- Use resource hints for critical resources
