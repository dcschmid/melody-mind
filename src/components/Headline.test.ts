/* eslint-disable max-lines-per-function */
import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { describe, expect, test } from "vitest";

import Headline from "../components/Headline.astro";

describe("Headline Component", () => {
  test("renders headline with title prop", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Test Headline",
        level: "h1",
      },
    });

    expect(result).toContain("<h1");
    expect(result).toContain("Test Headline");
    expect(result).toContain("headline");
  });

  test("renders headline with slot content", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      slots: {
        default: "Slot headline content",
      },
    });

    expect(result).toContain("Slot headline content");
    expect(result).toContain("<h1"); // Default level
  });

  test("renders different heading levels correctly", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "H2 Headline",
        level: "h2",
      },
    });

    expect(result).toContain("<h2");
    expect(result).toContain("H2 Headline");
    expect(result).toContain("</h2>");
  });

  test("applies custom CSS classes", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Custom Styled Headline",
        className: "headline--large custom-class",
      },
    });

    expect(result).toContain("headline--large custom-class");
    expect(result).toContain("Custom Styled Headline");
  });

  test("renders with custom ID", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Headline with ID",
        id: "custom-headline-id",
      },
    });

    expect(result).toContain('id="custom-headline-id"');
    expect(result).toContain("Headline with ID");
  });

  test("renders with aria-label", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Accessible Headline",
        ariaLabel: "Custom aria label",
      },
    });

    expect(result).toContain('aria-label="Custom aria label"');
    expect(result).toContain("Accessible Headline");
  });

  test("renders with focusable attributes when focusable is true", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Focusable Headline",
        focusable: true,
      },
    });

    expect(result).toContain('tabindex="-1"');
    expect(result).toContain("headline--focusable");
    expect(result).toContain("Focusable Headline");
  });

  test("prioritizes title prop over slot content", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Title prop content",
      },
      slots: {
        default: "Slot content",
      },
    });

    expect(result).toContain("Title prop content");
    expect(result).not.toContain("Slot content");
  });

  test("renders with default values when no props provided", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      slots: {
        default: "Default headline",
      },
    });

    expect(result).toContain("<h1"); // Default level
    expect(result).toContain("headline"); // Base BEM class
    expect(result).toContain("Default headline");
    expect(result).not.toContain('tabindex="-1"'); // Default focusable=false
  });

  test("renders all heading levels correctly", async () => {
    const container = await AstroContainer.create();
    const levels = ["h1", "h2", "h3", "h4", "h5", "h6"] as const;

    for (const level of levels) {
      const result = await container.renderToString(Headline, {
        props: {
          title: `${level} Headline`,
          level,
        },
      });

      expect(result).toContain(`<${level}`);
      expect(result).toContain(`</${level}>`);
      expect(result).toContain(`${level} Headline`);
    }
  });

  test("renders with focusable=false explicitly", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Non-focusable Headline",
        focusable: false,
      },
    });

    expect(result).not.toContain('tabindex="-1"');
    expect(result).not.toContain("headline--focusable");
    expect(result).toContain("headline");
    expect(result).toContain("Non-focusable Headline");
  });

  test("handles empty title prop gracefully", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "",
      },
      slots: {
        default: "Fallback content",
      },
    });

    // Empty title should fall back to slot content
    expect(result).toContain("Fallback content");
  });

  test("renders with combination of multiple props", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Complex Headline",
        level: "h3",
        className: "headline--large custom-style",
        id: "complex-headline",
        focusable: true,
        ariaLabel: "Complex headline description",
      },
    });

    expect(result).toContain("<h3");
    expect(result).toContain("Complex Headline");
    expect(result).toContain("headline--large custom-style");
    expect(result).toContain('id="complex-headline"');
    expect(result).toContain('tabindex="-1"');
    expect(result).toContain('aria-label="Complex headline description"');
    expect(result).toContain("headline--focusable");
  });

  test("includes WCAG AAA compliant styling", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline, {
      props: {
        title: "Accessible Headline",
      },
    });

    // Verify BEM base class is present
    expect(result).toContain("headline");
    expect(result).toContain("Accessible Headline");
    // CSS styles are applied via stylesheet, not inline classes
    expect(result).toContain('class="headline"');
  });

  test("handles missing slot content gracefully", async () => {
    const container = await AstroContainer.create();
    const result = await container.renderToString(Headline);

    expect(result).toContain("<h1"); // Should still render the tag
    expect(result).toContain("</h1>");
    // Content should be empty but structure intact
  });

  test("renders focus ring styles only when focusable", async () => {
    const container = await AstroContainer.create();

    // Test focusable version
    const focusableResult = await container.renderToString(Headline, {
      props: {
        title: "Focusable",
        focusable: true,
      },
    });

    expect(focusableResult).toContain("headline--focusable");

    // Test non-focusable version
    const nonFocusableResult = await container.renderToString(Headline, {
      props: {
        title: "Non-focusable",
        focusable: false,
      },
    });

    expect(nonFocusableResult).not.toContain("headline--focusable");
    expect(nonFocusableResult).toContain("headline");
  });

  // ======================================
  // SNAPSHOT TESTS
  // ======================================

  describe("Snapshot Tests", () => {
    test("matches snapshot for basic headline", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        props: {
          title: "Basic Headline",
        },
      });

      expect(result).toMatchSnapshot();
    });

    test("matches snapshot for headline with all props", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        props: {
          title: "Complete Headline",
          level: "h2",
          className: "headline--large custom-class",
          id: "complete-headline",
          focusable: true,
          ariaLabel: "Complete headline description",
        },
      });

      expect(result).toMatchSnapshot();
    });

    test("matches snapshot for different heading levels", async () => {
      const container = await AstroContainer.create();
      const levels = ["h1", "h2", "h3", "h4", "h5", "h6"] as const;

      for (const level of levels) {
        const result = await container.renderToString(Headline, {
          props: {
            title: `Level ${level} Headline`,
            level,
          },
        });

        expect(result).toMatchSnapshot(`headline-${level}`);
      }
    });

    test("matches snapshot for slot content", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        slots: {
          default: "Slot content headline",
        },
      });

      expect(result).toMatchSnapshot();
    });

    test("matches snapshot for empty content", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline);

      expect(result).toMatchSnapshot();
    });
  });

  // ======================================
  // WCAG 2.2 AAA ACCESSIBILITY TESTS
  // ======================================

  describe("WCAG 2.2 AAA Accessibility Tests", () => {
    test("ensures proper semantic heading structure", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        props: {
          title: "Semantic Test",
          level: "h2",
        },
      });

      // Must use semantic HTML heading elements
      expect(result).toContain("<h2");
      expect(result).toContain("</h2>");
      expect(result).not.toContain('<div role="heading"'); // Avoid role="heading" when semantic elements available
    });

    test("provides accessible focus management", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        props: {
          title: "Focus Test",
          focusable: true,
        },
      });

      // WCAG: Focus must be visible and manageable
      expect(result).toContain('tabindex="-1"'); // Programmatically focusable
      expect(result).toContain("headline--focusable"); // CSS class for focus styles
    });

    test("supports screen reader accessibility", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        props: {
          title: "Screen Reader Test",
          ariaLabel: "Detailed description for screen readers",
        },
      });

      // WCAG: Provide alternative text descriptions
      expect(result).toContain('aria-label="Detailed description for screen readers"');
      expect(result).toContain("Screen Reader Test"); // Visible text should still be present
    });

    test("maintains proper document structure with IDs", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        props: {
          title: "Document Structure Test",
          id: "unique-section-heading",
          level: "h3",
        },
      });

      // WCAG: Support skip navigation and document outline
      expect(result).toContain('id="unique-section-heading"');
      expect(result).toContain("<h3"); // Proper heading level for document hierarchy
    });

    test("handles text overflow accessibly", async () => {
      const container = await AstroContainer.create();
      const longTitle = "A".repeat(200); // Very long title to test text wrapping
      const result = await container.renderToString(Headline, {
        props: {
          title: longTitle,
        },
      });

      // WCAG: Text must be readable and not hidden
      expect(result).toContain(longTitle);
      expect(result).toContain("headline"); // CSS class includes text-wrap and word-break
    });

    test("supports keyboard navigation requirements", async () => {
      const container = await AstroContainer.create();

      // Test non-focusable (standard heading)
      const standardResult = await container.renderToString(Headline, {
        props: {
          title: "Standard Heading",
          focusable: false,
        },
      });

      // Standard headings should not be in tab order
      expect(standardResult).not.toContain("tabindex");

      // Test focusable (for skip navigation)
      const focusableResult = await container.renderToString(Headline, {
        props: {
          title: "Skip Target Heading",
          focusable: true,
        },
      });

      // Skip navigation targets should be programmatically focusable
      expect(focusableResult).toContain('tabindex="-1"');
    });

    test("ensures consistent heading hierarchy", async () => {
      const container = await AstroContainer.create();
      const levels = ["h1", "h2", "h3", "h4", "h5", "h6"] as const;

      for (const level of levels) {
        const result = await container.renderToString(Headline, {
          props: {
            title: `${level.toUpperCase()} Content`,
            level,
          },
        });

        // WCAG: Heading hierarchy must be logical and consistent
        expect(result).toContain(`<${level}`);
        expect(result).toContain(`</${level}>`);
        expect(result).not.toContain('<div role="heading"'); // Use semantic elements
      }
    });

    test("maintains content accessibility with fallbacks", async () => {
      const container = await AstroContainer.create();

      // Test with both title and slot (title should take precedence)
      const bothContentResult = await container.renderToString(Headline, {
        props: {
          title: "Title Priority",
        },
        slots: {
          default: "Slot Fallback",
        },
      });

      // WCAG: Content must be predictable and accessible
      expect(bothContentResult).toContain("Title Priority");
      expect(bothContentResult).not.toContain("Slot Fallback");

      // Test with only slot content
      const slotOnlyResult = await container.renderToString(Headline, {
        slots: {
          default: "Slot Content Only",
        },
      });

      expect(slotOnlyResult).toContain("Slot Content Only");
    });

    test("provides proper ARIA labeling support", async () => {
      const container = await AstroContainer.create();

      // Test with aria-label
      const ariaLabelResult = await container.renderToString(Headline, {
        props: {
          title: "Visible Text",
          ariaLabel: "Detailed accessible name",
        },
      });

      // WCAG: ARIA label should supplement, not replace visible text
      expect(ariaLabelResult).toContain('aria-label="Detailed accessible name"');
      expect(ariaLabelResult).toContain("Visible Text");

      // Test without aria-label
      const noAriaResult = await container.renderToString(Headline, {
        props: {
          title: "Standard Heading",
        },
      });

      expect(noAriaResult).not.toContain("aria-label");
      expect(noAriaResult).toContain("Standard Heading");
    });

    test("ensures CSS classes support accessibility features", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        props: {
          title: "Accessibility CSS Test",
          focusable: true,
          className: "headline--large",
        },
      });

      // WCAG: CSS classes must support accessibility features
      expect(result).toContain("headline"); // Base accessibility styles
      expect(result).toContain("headline--focusable"); // Focus indicator styles
      expect(result).toContain("headline--large"); // Size modifier preserved

      // Classes should be properly combined
      expect(result).toMatch(/class="[^"]*headline[^"]*"/);
      expect(result).toMatch(/class="[^"]*headline--large[^"]*"/);
      expect(result).toMatch(/class="[^"]*headline--focusable[^"]*"/);
    });

    test("validates HTML structure integrity", async () => {
      const container = await AstroContainer.create();
      const result = await container.renderToString(Headline, {
        props: {
          title: "HTML Structure Test",
          level: "h4",
          id: "structure-test",
          className: "custom-class",
          focusable: true,
          ariaLabel: "Structure validation",
        },
      });

      // WCAG: HTML must be valid and well-formed
      expect(result).toMatch(/<h4[^>]*>.*<\/h4>/); // Proper tag structure
      expect(result).toMatch(/id="structure-test"/); // Valid ID attribute
      expect(result).toMatch(/class="[^"]*"/); // Valid class attribute
      expect(result).toMatch(/aria-label="[^"]*"/); // Valid ARIA attribute
      expect(result).toMatch(/tabindex="-1"/); // Valid tabindex

      // Should not contain malformed attributes
      expect(result).not.toContain('=""'); // No empty attributes
      expect(result).not.toContain("undefined"); // No undefined values
      expect(result).not.toContain("null"); // No null values
    });
  });
});
