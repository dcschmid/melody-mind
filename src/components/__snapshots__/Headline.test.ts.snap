// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Headline Component > Snapshot Tests > matches snapshot for basic headline 1`] = `"<h1 class="headline" data-astro-cid-odmoivsz="true">Basic Headline</h1>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for different heading levels > headline-h1 1`] = `"<h1 class="headline" data-astro-cid-odmoivsz="true">Level h1 Headline</h1>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for different heading levels > headline-h2 1`] = `"<h2 class="headline" data-astro-cid-odmoivsz="true">Level h2 Headline</h2>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for different heading levels > headline-h3 1`] = `"<h3 class="headline" data-astro-cid-odmoivsz="true">Level h3 Headline</h3>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for different heading levels > headline-h4 1`] = `"<h4 class="headline" data-astro-cid-odmoivsz="true">Level h4 Headline</h4>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for different heading levels > headline-h5 1`] = `"<h5 class="headline" data-astro-cid-odmoivsz="true">Level h5 Headline</h5>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for different heading levels > headline-h6 1`] = `"<h6 class="headline" data-astro-cid-odmoivsz="true">Level h6 Headline</h6>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for empty content 1`] = `"<h1 class="headline" data-astro-cid-odmoivsz="true"></h1>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for headline with all props 1`] = `"<h2 id="complete-headline" aria-label="Complete headline description" tabindex="-1" class="headline headline--large custom-class headline--focusable" data-astro-cid-odmoivsz="true">Complete Headline</h2>"`;

exports[`Headline Component > Snapshot Tests > matches snapshot for slot content 1`] = `"<h1 class="headline" data-astro-cid-odmoivsz="true">Slot content headline</h1>"`;
